api/envoy/extensions/wasm/v3/wasm.proto

```diff
@@ -17,6 +18,23 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: Wasm]
 // [#extension: envoy.bootstrap.wasm]
 
+// If there is a fatal error on the VM (e.g. exception, abort()), then the policy will be applied.
+enum FailurePolicy {
+  // No policy is specified. The default policy will be used. The default policy is ``FAIL_RELOAD``.
+  UNSPECIFIED = 0;
+
+  // All plugins associated with the VM will be reloaded then new requests are processed. This
+  // make sense when the VM failure is caused by runtime exception, abort(), or panic().
+  FAIL_RELOAD = 1;
```

Comment:
Is there any backoff applied here? What happens if the VM fails again as soon as it's started? Could this be a vector for an attacker to DoS the data plane by forcing it to continously reload the VM? I'm thinking that this might magnify a query-of-death problem in unpleasant ways.

Reply:
How about a reload rate limit support like allow 2 mins 24 failures and make it configurable? Then this should be enough to address the occasional crash and won't bring new risk?

Reply:
It might be easiest to just use the same exponential backoff computation that we use for retries.

We also need to define what happens to new requests while we're in backoff delay.  We might want to do something like this:

```
// What to do with requests when the VM fails.
// (Maybe this can just remain a bool rather than being an enum?)
enum FailurePolicy {
  UNSPECIFIED = 0;
  FAIL_CLOSED = 1;
  FAIL_OPEN = 2;
}
FailurePolicy failure_policy = 7;

message ReloadBehavior {
  // Parameters to control backoff delay before reloading.
  // (not sure exactly what we want here)
  google.protobuf.Duration min_backoff = 1;
  google.protobuf.Duration max_backoff = 2;

  enum Policy {
    UNSPECIFIED = 0;
    // Requests will be queued while in backoff.
    RELOAD_WAIT = 1;
    // Requests will not be held while in backoff.
    // Instead, the FailurePolicy will be applied.
    RELOAD_NO_WAIT = 2;
  }
  Policy policy = 3;
}
// If unset, VMs are not reloaded upon failure.
ReloadBehavior reload_behavior = 8;
```

That way, we can configure whether to reload independently from whether we fail open or closed.

Reply:
Wasm VM is supposed to be snapshot after compilation and the configuration callback (`onConfigure`) which should run on the main thread during xDS processing. After that, making a copy should be cheap, and can be done per worker. This is definitely worth verifying.

Reply:
> Actually, doesn't the reload itself require some I/O, like re-reading the plugin from disk? [...] I am assuming that the plugin is shared across all requests being processed by the filter, so reloading the plugin is not just about how to handle the current request but actually has implications for subsequent requests as well

Just to elaborate on what @kyessenov said, the loading of plugin bytes occurs during xDS processing, and then that "base" WasmVM is cloned to create the per-worker-thread Wasm VMs that run for processing requests. So e.g. failure of one WasmVM wouldn't by itself interfere with Wasm execution in other worker threads. However, if the crash is due to some bug in plugin logic, then it's likely to surface in other WasmVMs as well.

Reply:
> Does the reloading have to happen when we create the filter chain? Couldn't we instead do it when the request gets to the filter?

I choose to reload the VM when we creating the filter chain instance because:

1. WASM is used for both L7 filters, L4 filters, Logger. If we treat it when the request gets to the filters, then we need to add the restart logic to multiple different positions.
2. Keep the restart logic be separated with the normal filter logic.

> Actually, doesn't the reload itself require some I/O, like re-reading the plugin from disk? If so, doesn't this need to be an async operation anyway, which would imply that the request needs to wait for it to complete anyway?

The WASM will be loaded when the xds loading the configuration. And restart/reload only create a clone from the exist base WASM vm. So, there basically has no additional network/disk IO. Async make no much sense for this scenario.

>  Also, why does FAIL_RELOAD always fall back to FAIL_CLOSED? What if we want to trigger a reload on runtime errors but fail open for other errors?

It's easy to extend `ReloadConfig` to provide other fallbacks. But it needn't be part of this PR. (I choose FAIL_CLOSED as default fallback because the `FAIL_CLOSED` is our default failure policy anyway.)